src/sdam/monitor.rs

```diff
@@ -184,7 +189,45 @@ impl HeartbeatMonitor {
         let result = match self.connection {
             Some(ref mut conn) => {
                 let command = is_master_command(self.client_options.server_api.as_ref());
-                run_is_master(command, conn).await
+                if let Some(ref handler) = self.client_options.sdam_event_handler {
+                    if self.topology.is_alive() {
+                        let event = ServerHeartbeatStartedEvent {
+                            server_address: self.address.clone(),
+                        };
+                        handler.handle_server_heartbeat_started_event(event);
+                    }
+                }
+                let is_master_result = run_is_master(command, conn).await;
```

Comment:
I'm starting to think using the `run_is_master` function here is more trouble than its worth. If we were to inline the contents of it here, we could use the actual command response from the server in the `reply` field rather than round tripping through our struct, which currently may accidentally discard fields (the `ok` field being an example). It would also avoid the optional duration stuff I mentioned in my other comment.

I ran into something similar when implementing the raw BSON stuff in the executor and ended up storing the server response and the deserialized one separately, allowing the event handler to use the unmodified server resposne while the driver logic used the deserialized one.